mutiny-core/src/labels.rs

```diff
@@ -0,0 +1,492 @@
+                Some(mut label_item) => {
+                    // Add the address to the label item
+                    // and sort so we can dedup the addresses
+                    label_item.addresses.push(address.clone());
+                    label_item.addresses.sort();
+                    label_item.addresses.dedup();
+
+                    // Update the last used timestamp
+                    label_item.last_used = now;
+
+                    self.set(key, label_item)?;
+                }
+                None => {
+                    // Create a new label item
+                    let label_item = LabelItem {
+                        addresses: vec![address.clone()],
+                        last_used: now,
+                    };
+                    self.set(key, label_item)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_contacts(&self) -> Result<HashMap<String, Contact>, MutinyError> {
+        let all = self.scan(CONTACT_PREFIX, None)?;
+        // remove the prefix from the keys
+        let mut contacts = HashMap::new();
+        for (key, contact) in all {
+            let label = key.replace(CONTACT_PREFIX, "");
+            contacts.insert(label, contact);
+        }
+
+        Ok(contacts)
+    }
+
+    fn get_contact(&self, label: impl AsRef<str>) -> Result<Option<Contact>, MutinyError> {
+        self.get(get_contact_key(label))
+    }
+
+    fn create_contact_from_label(
+        &self,
+        label: impl AsRef<str>,
+        contact: Contact,
+    ) -> Result<(), MutinyError> {
+        let key = get_contact_key(label);
+        self.set(key, contact)?;
+        Ok(())
+    }
+
+    /// Create a new contact and return the identifying label
+    fn create_new_contact(&self, contact: Contact) -> Result<String, MutinyError> {
+        // generate a uuid, this will be the "label" that we use to store the contact
+        let id = Uuid::new_v4().to_string();
+        let key = get_contact_key(&id);
+        self.set(key, contact)?;
+        Ok(id)
+    }
+
+    fn get_tag_items(&self) -> Result<Vec<TagItem>, MutinyError> {
+        let mut tag_items = vec![];
+
+        // Get all the contacts
+        let mut contacts = self.get_contacts()?;
+        // Get all the labels
+        let mut labels = self.get_labels()?;
+
+        // filter out contacts that have a label
+        contacts.retain(|label, _| labels.get(label).is_none());
+        // filter out labels that have a contact
+        labels.retain(|label, _| contacts.get(label).is_none());
+
+        // Convert the labels into tag items
+        tag_items.extend(
+            labels
+                .into_iter()
+                .map(|(label, label_item)| TagItem::Label((label, label_item))),
+        );
+
+        // Convert the contacts into tag items
+        tag_items.extend(
+            contacts
+                .into_iter()
+                .map(|(id, c)| TagItem::Contact((id, c))),
+        );
```

Comment:
>  How can a label, which is only supposed to reference an on chain address, also be a contact

Like I said below but a contact is only useful if is a label so it can be referenced throughout a list of transactions so users can see where funds are coming from/going to. Contacts use labels as their identifiers so we can get all the nice features of labels without having to re-implement it.

> How is a contact an on chain address?

That never happens.